src/cargo/ops/registry/publish.rs

```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
+    let just_pkgs: Vec<_> = pkgs.iter().map(|p| p.0).collect();
+    let reg_or_index = match opts.reg_or_index.clone() {
+        Some(r) => {
+            validate_registry(&just_pkgs, Some(&r))?;
+            Some(r)
+        }
+        None => {
+            let reg = super::infer_registry(&just_pkgs)?;
+            validate_registry(&just_pkgs, reg.as_ref())?;
+            if let Some(RegistryOrIndex::Registry(ref registry)) = &reg {
+                if registry != CRATES_IO_REGISTRY {
+                    // Don't warn for crates.io.
+                    opts.gctx.shell().note(&format!(
+                        "found `{}` as only allowed registry. Publishing to it automatically.",
+                        registry
+                    ))?;
+                }
             }
+            reg
         }
+    };
 
-        let reg_name = publish_registry
-            .clone()
-            .unwrap_or_else(|| CRATES_IO_REGISTRY.to_string());
-        if allowed_registries.is_empty() {
-            bail!(
-                "`{}` cannot be published.\n\
-                 `package.publish` must be set to `true` or a non-empty list in Cargo.toml to publish.",
-                pkg.name(),
-            );
-        } else if !allowed_registries.contains(&reg_name) {
-            bail!(
-                "`{}` cannot be published.\n\
-                 The registry `{}` is not listed in the `package.publish` value in Cargo.toml.",
-                pkg.name(),
-                reg_name
-            );
-        }
-    }
     // This is only used to confirm that we can create a token before we build the package.
     // This causes the credential provider to be called an extra time, but keeps the same order of errors.
-    let ver = pkg.version().to_string();
-    let operation = Operation::Read;
-
-    let reg_or_index = match opts.reg_or_index.clone() {
-        Some(RegistryOrIndex::Registry(_)) | None => {
-            publish_registry.map(RegistryOrIndex::Registry)
-        }
-        val => val,
-    };
     let source_ids = super::get_source_id(opts.gctx, reg_or_index.as_ref())?;
     let mut registry = super::registry(
         opts.gctx,
         &source_ids,
         opts.token.as_ref().map(Secret::as_deref),
         reg_or_index.as_ref(),
         true,
-        Some(operation).filter(|_| !opts.dry_run),
+        Some(Operation::Read).filter(|_| !opts.dry_run),
     )?;
-    verify_dependencies(pkg, &registry, source_ids.original)?;
 
-    // Prepare a tarball, with a non-suppressible warning if metadata
-    // is missing since this is being put online.
-    let tarball = ops::package_one(
+    // Validate all the packages before publishing any of them.
+    for (pkg, _) in &pkgs {
+        verify_dependencies(pkg, &registry, source_ids.original)?;
+    }
+
+    let specs = Packages::from_flags(
+        false,
+        vec![],
+        just_pkgs.iter().map(|p| p.name().to_string()).collect(),
```

Comment:
package names are not always sufficient specs.  I don't know what paths this code goes down but if it hits the wrong one now, or in the future, this will break in less commonly seen but important cases.

For example, `clap` has transitive dependencies on its self so I can't run `cargo check -p clap`, I have to do `cargo check -p clap@...`